src/configuration/action.rs

```diff
@@ -13,12 +14,24 @@ pub struct Action {
     #[serde(default)]
     pub conditions: Vec<PatternExpression>,
     #[serde(default)]
+    pub predicates: Vec<String>,
+    #[serde(skip_deserializing)]
+    pub compiled_predicates: OnceCell<Vec<Predicate>>,
+    #[serde(default)]
     pub data: Vec<DataItem>,
 }
 
 impl Action {
     pub fn conditions_apply(&self) -> bool {
-        self.conditions.is_empty() || self.conditions.iter().all(|m| m.applies())
+        let predicates = self
+            .compiled_predicates
+            .get()
+            .expect("predicates must be compiled by now");
+        if predicates.is_empty() {
```

Comment:
we need to document clearly that `predicates` take precedence and when at least one predicate is being defined, conditions are not even being evaluated.

Reply:
We can... if this does NOT 🔥 by the release.